### PR TITLE
Fix handling of unset values

### PR DESCRIPTION
**Pull request by pcmanus**
In C*/DSE, the unset value is implement through a specific `ByteBuffer` instance that is `public static` (and reference equality is used on that instance, so we need _that_ instance). We cannot however access directly (as a static variable) from the stargate code.

The current code (technically, the `CBUtil` class that is used to deserialize client queries) was actually using the `ByteBuffer#UNSET_BYTE_BUFFER` from the C* jar that the `cql` module import, which is definitively not the one the persistence layer expects.

I'll note that there is 2 options to fix this:
1. we define our own `public static` ByteBuffer instance (in say `Persistence`), use that from the `cql` module (and any other consumer for the `Persistence` API that needs it), and require each persistence extensions to translate that value into their own instance. The main downside of this approach is:
    - every persistence extensions has to walk over every value it gets to check for the 'stargate unset' and replace it by its own 'internal unset'.
2. we add a new method (`unsetValue()`) to the `Persistence` API so that each persistence extension exposes the actual instance it uses for unset values. We can then use that value in the `cql` module. The downsides of this approach are:
    - it's one more method to the `Persistence` API (albeit a trivial one to implement).
    - you need access to the persistence to be able to produce an unset value.  Which is not a huge deal, except that the code that currently produce those, `CBUtil` is all static, so the patch currently uses an ugly hack to work around that.

This PR actually implement the 2nd option. My rational is that having to convert the values within the persistence layer feels error prone (easy to forget an entry point) and requires more code, code that ends up duplicated for every persistence extensions. The downsides of option 2 feel smaller.  That is, the hack in `CBUtil` is admittedly ugly, but it's very small and can be refactored out at some point (and we need to give some love to the `cql` module at some point, at least to remove the dependencies on `cassandra-all` it currently have).

### CODE REVIEW

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- testing/src/test/java/io/stargate/it/CQLTest.java  3
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java  -2
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6267779&bid=20405960)

File: cql/src/main/java/org/apache/cassandra/stargate/transport/internal/CBUtil.java [hunk 1]
```diff
@@ -367,7 +391,7 @@ public static ByteBuffer readBoundValue(ByteBuf cb, ProtocolVersion protocolVers
           ProtocolVersion.V4)) // backward compatibility for pre-version 4
       return null;
       if (length == -1) return null;
-      else if (length == -2) return ByteBufferUtil.UNSET_BYTE_BUFFER;
+      else if (length == -2) return unsetValue;
```
Comment (mpenick) [inline cql/src/main/java/org/apache/cassandra/stargate/transport/internal/CBUtil.java:394]: minor: Maybe we should assert that `unsetValue` is set?

File: persistence-api/src/main/java/io/stargate/db/Persistence.java [hunk 2]
```diff
@@ -58,6 +58,16 @@
 
   DataStore newDataStore(QueryState<Q> state, QueryOptions queryOptions);
 
+  /**
+   * The object that should be used to act as an 'unset' value for this persistence (for use in
+   * {@link QueryOptions#getValues()} or in the {@code values} argument of {@link #batch}).
+   *
+   * <p>Please note that persistence implementations are allowed to use <b>reference equality</b> to
+   * detect this value, so the object returned by this method should be used "as-is" and should
+   * <b>not</b> be copied (through {@link ByteBuffer#duplicate()} or any other method).
```
Comment (mpenick) [inline persistence-api/src/main/java/io/stargate/db/Persistence.java:67]: Awesome comments!

File: cql/src/main/java/org/apache/cassandra/stargate/transport/internal/CBUtil.java (hunk 1 above)
Comment (pcmanus) [reply cql/src/main/java/org/apache/cassandra/stargate/transport/internal/CBUtil.java:394]: Yes. Meant to but forgot, sorry.

File: testing/src/test/java/io/stargate/it/CQLTest.java [hunk 3]
```diff
@@ -187,59 +189,126 @@ public void querySystemLocal() throws UnknownHostException {
...
 
   @Test
-  public void querySimpleBidingMarkersByName() {
-    querySimple(true);
+  public void querySimpleBindingByName() {
+    querySimple(false, true);
   }
 
-  private void querySimple(boolean bindMarkersByName) {
+  private void querySimple(boolean usePrepared, boolean bindByName) {
```
Comment (dimas-b) [inline testing/src/test/java/io/stargate/it/CQLTest.java:253]: nit: with JUnit 5 we could convert this method to a `@ParameterizedTest` with `@CsvSource` for example.

Comment (dimas-b) [reply testing/src/test/java/io/stargate/it/CQLTest.java:253]: also `unsetValues`

**Review by dimas-b — DISMISSED** [review]

**Comment by dimas-b** [comment]
C* 4.0 integration tests have been merged into `master`. Please re-run CI :)

File: testing/src/test/java/io/stargate/it/CQLTest.java (hunk 3 above)
Comment (pcmanus) [reply testing/src/test/java/io/stargate/it/CQLTest.java]: Good idea. I ended up using an enum and `@EnumSource`. I think that's a tad more readable here.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- testing/src/test/java/io/stargate/it/CQLTest.java  3
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java  -2
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6267779&bid=20405960)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity increasing per file
==============================
- testing/src/test/java/io/stargate/it/CQLTest.java  3
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java  -2
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6267779&bid=20405960)

**Review by jorgebay — APPROVED** [review]
Good effort of documenting it in the code 👍

**Review by mpenick — APPROVED** [review]

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity increasing per file
==============================
- testing/src/test/java/io/stargate/it/CQLTest.java  3
         

Clones removed
==============
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java  -2
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6267779&bid=20405960)

**Pull request merged**